data_pipeline_api/java_api/src/main/java/uk/ramp/file/CleanableFileChannel.java

```diff
@@ -0,0 +1,150 @@
+package uk.ramp.file;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.ref.Cleaner;
+import java.lang.ref.Cleaner.Cleanable;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileChannel.MapMode;
+import java.nio.channels.FileLock;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.ScatteringByteChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * Wrapper around standard FileChannel with additional safety net for closing if user does not close
+ * and ability to run custom function during cleanup/closing. Users should either use with
+ * try-with-resources or ensure they manually call .close() once finished. As a safety net, .close()
+ * will be called by a cleaner if and when the instance of CleanableFileChannel is going to be
+ * collected by GC.
+ */
+public class CleanableFileChannel
+    implements AutoCloseable, SeekableByteChannel, GatheringByteChannel, ScatteringByteChannel {
+  private static final Cleaner cleaner = Cleaner.create(); // safety net for closing
+  private final Cleanable cleanable;
+
+  private final FileChannelWrapper fileChannelWrapper;
+
+  public CleanableFileChannel(FileChannel fileChannel, Runnable runOnClose) {
+    this.fileChannelWrapper = new FileChannelWrapper(fileChannel, runOnClose);
+    this.cleanable = cleaner.register(this, fileChannelWrapper);
+  }
+
+  // Defining a resource that requires cleaning
+  private static class FileChannelWrapper implements Runnable {
+    private final FileChannel fileChannel;
+    private final Runnable runOnClose;
+
+    FileChannelWrapper(FileChannel fileChannel, Runnable runOnClose) {
+      this.fileChannel = fileChannel;
+      this.runOnClose = runOnClose;
+    }
+
+    // Invoked by close method or cleaner
+    @Override
+    public void run() {
+      runOnClose.run();
+      try {
+        fileChannel.close();
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    return fileChannelWrapper.fileChannel.read(dst);
+  }
+
+  @Override
+  public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+    return fileChannelWrapper.fileChannel.read(dsts, offset, length);
+  }
+
+  @Override
+  public long read(ByteBuffer[] dsts) throws IOException {
+    return 0;
```

Comment:
This looks a bit different to all the other overridden methods.  Why is this not doing a read?

Reply:
Thanks for spotting this mistake. I've updated it.